controllers/templatesync/template_sync.go

```diff
@@ -854,51 +866,29 @@ func (r *PolicyReconciler) cleanUpExcessTemplates(
 
 	tmplGVRs := []gvrScoped{}
 
-	// Query for ConstraintTemplates and collect the GroupVersionResource for each Constraint
-	gkConstraintTemplateListv1 := gktemplatesv1.ConstraintTemplateList{}
+	// Query for Constraints if one was already synced successfully or the boolean is not yet set
+	if r.createdGkConstraint == nil || *r.createdGkConstraint {
+		// Query for ConstraintTemplates and collect the GroupVersionResource for each Constraint
+		gkConstraintTemplateListv1 := gktemplatesv1.ConstraintTemplateList{}
 
-	err := r.List(ctx, &gkConstraintTemplateListv1, &client.ListOptions{})
-	if err == nil {
-		// Add the ConstraintTemplate to the GVR list
-		if len(gkConstraintTemplateListv1.Items) > 0 {
-			tmplGVRs = append(tmplGVRs, gvrScoped{
-				gvr: schema.GroupVersionResource{
-					Group:    utils.GvkConstraintTemplate.Group,
-					Resource: "constrainttemplates",
-					Version:  "v1",
-				},
-				namespaced: false,
-			})
-		}
-		// Iterate over the ConstraintTemplates to gather the Constraints on the cluster
-		for _, gkCT := range gkConstraintTemplateListv1.Items {
-			tmplGVRs = append(tmplGVRs, gvrScoped{
-				gvr: schema.GroupVersionResource{
-					Group:    utils.GConstraint,
-					Resource: strings.ToLower(gkCT.Spec.CRD.Spec.Names.Kind),
-					Version:  "v1beta1",
-				},
-				namespaced: false,
-			})
-		}
-	} else if meta.IsNoMatchError(err) {
-		// If there's no v1 ConstraintTemplate, try the v1beta1 version
-		gkConstraintTemplateListv1beta1 := gktemplatesv1beta1.ConstraintTemplateList{}
-		err := r.List(ctx, &gkConstraintTemplateListv1beta1, &client.ListOptions{})
+		err := r.List(ctx, &gkConstraintTemplateListv1, &client.ListOptions{})
 		if err == nil {
 			// Add the ConstraintTemplate to the GVR list
 			if len(gkConstraintTemplateListv1.Items) > 0 {
 				tmplGVRs = append(tmplGVRs, gvrScoped{
 					gvr: schema.GroupVersionResource{
 						Group:    utils.GvkConstraintTemplate.Group,
 						Resource: "constrainttemplates",
-						Version:  "v1beta1",
+						Version:  "v1",
 					},
 					namespaced: false,
 				})
+			} else {
+				// No ConstraintTemplates found -- reset boolean
+				*r.createdGkConstraint = false
```

Comment:
You'll want to explicitly set this to `false` on errors when the CRD doesn't exist error is encountered.